src/central-server/admin-service/core/src/main/java/org/niis/xroad/cs/admin/core/service/ConfigurationServiceImpl.java

```diff
@@ -94,6 +129,69 @@ public ConfigurationAnchor getConfigurationAnchor(String sourceType) {
         return new ConfigurationAnchor(configurationSource.getAnchorFileHash(), configurationSource.getAnchorGeneratedAt());
     }
 
+    @Override
+    public ConfigurationAnchor recreateAnchor(String configurationType) {
+        auditEventHelper.changeRequestScopedEvent(configurationType.equals(INTERNAL_CONFIGURATION)
+                ? RE_CREATE_INTERNAL_CONFIGURATION_ANCHOR
+                : RE_CREATE_EXTERNAL_CONFIGURATION_ANCHOR);
+
+        final var instanceIdentifier = Optional.ofNullable(systemParameterService.getInstanceIdentifier())
+                .filter(StringUtils::isNotEmpty)
+                .orElseThrow(() -> new ConfigurationSourceException(INSTANCE_IDENTIFIER_NOT_SET));
+
+        final var configurationSource = configurationSourceRepository.findBySourceType(configurationType.toLowerCase())
+                .orElseThrow(ConfigurationServiceImpl::notFoundException);
+
+        if (CollectionUtils.isEmpty(configurationSource.getConfigurationSigningKeys())) {
+            throw new ConfigurationSourceException(NO_CONFIGURATION_SIGNING_KEYS_CONFIGURED);
+        }
+
+        final var sources = configurationSourceRepository.findAllBySourceType(configurationType.toLowerCase());
+        final var now = ZonedDateTime.now(ZoneId.of("UTC"));
+        final var anchorXml = buildAnchorXml(configurationType, instanceIdentifier, now, sources);
+        final var anchorXmlBytes = anchorXml.getBytes(StandardCharsets.UTF_8);
+        final var anchorXmlHash = calculateAnchorHexHash(anchorXmlBytes);
+        for (final var src : sources) {
+            if (src.getConfigurationSigningKey() != null) {
+                src.setAnchorGeneratedAt(now.toInstant());
+                src.setAnchorFileHash(anchorXmlHash);
+                src.setAnchorFile(anchorXmlBytes);
+                configurationSourceRepository.save(src);
+            }
+        }
+
+        return new ConfigurationAnchor(anchorXmlHash, now.toInstant());
+    }
+
+    private String buildAnchorXml(final String configurationType,
+                                  final String instanceIdentifier,
+                                  final ZonedDateTime now,
+                                  final List<ConfigurationSourceEntity> sources) {
+        try {
+            JAXBContext jaxbCtx = JAXBContext.newInstance(ObjectFactory.class);
```

Comment:
Can be declared on class level and initialized once (it is a resource intensive operation)